src/backward_reach.rs

```diff
@@ -0,0 +1,64 @@
+use std::fmt::Debug;
+
+pub trait PredBasis<T> {
+    fn pred_basis(&self, state: &T) -> Vec<T>;
+}
+
+fn exists_le_than<T: PartialOrd + Eq>(items: &Vec<T>, value: &T) -> bool {
+    items.into_iter().any(|item| item <= value)
+}
+
+fn exists_ge_than<T: PartialOrd + Eq>(items: &Vec<T>, value: &T) -> bool {
+    items.into_iter().any(|item| item >= value)
+}
+
+pub fn is_backward_reach<T: PartialOrd + Eq + Debug>(ts: &dyn PredBasis<T>, fin: T, init: Vec<T>) -> bool {
+    let mut to_explore: Vec<T> = vec![fin];
+    let mut explored: Vec<T> = Vec::new();
+
+    while to_explore.len() > 0 {
+        let curr = to_explore.pop().unwrap();
```

Comment:
Error management? I don't like unwrap without an explanation.

Better yet - you have a great explanation for that unwrap! So great that you can encode it in the code itself by changing the loop condition:

```suggestion
        while let Some(curr) = to_explore.pop() { ... }
```

[similar loop](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=4ef604fcf8dc53f539f35a51be84523e) which works thanks to a smart borrow checker which understand [Non Lexical Lifetimes (NLL)](https://rust-lang.github.io/rfcs/2094-nll.html).